src/dc_tools.rs

```diff
@@ -670,13 +675,114 @@ pub fn remove_subject_prefix(last_subject: &str) -> String {
         .to_string()
 }
 
+// Types and methods to create hop-info for message-info
+
+fn extract_address_from_receive_header<'a>(header: &'a str, start: &str) -> Option<&'a str> {
+    let header_len = header.len();
+    header.find(start).and_then(|mut begin| {
+        begin += start.len();
+        let end = header
+            .get(begin..)?
+            .find(|c| c == ' ' || c == '\n')
+            .unwrap_or(header_len);
+        header.get(begin..begin + end)
+    })
+}
+
+pub(crate) fn parse_receive_header(header: &str) -> String {
+    let mut hop_info = String::from("Hop:\n");
+
+    if let Ok(date) = dateparse(header) {
+        let date_obj = Local.timestamp(date, 0);
+        hop_info.push_str(&format!("Date: {}\n", date_obj.to_rfc2822()));
+    };
+
+    if let Some(from) = extract_address_from_receive_header(header, "from ") {
+        hop_info.push_str(&format!("From: {}\n", from.trim()));
+    }
+
+    if let Some(by) = extract_address_from_receive_header(header, "by ") {
+        hop_info.push_str(&format!("By: {}\n", by.trim()));
+    }
+    hop_info
+}
+
+/// parses "receive"-headers
+pub(crate) fn parse_receive_headers(headers: &Headers) -> String {
+    let headers = headers
+        .get_all_headers("Received")
+        .iter()
+        .rev()
+        .filter_map(|header_map_item| from_utf8(header_map_item.get_value_raw()).ok())
+        .enumerate()
+        .map(|(i, header_value)| (i + 1).to_string() + ". " + &parse_receive_header(header_value))
+        .collect::<Vec<_>>();
+
+    headers.iter().map(|a| a.to_string()).join("\n")
```

Comment:
This can be shortened:

```suggestion
    let headers = headers.get_all_headers("Received");
    headers
        .iter()
        .rev()
        .filter_map(|header_map_item| from_utf8(header_map_item.get_value_raw()).ok())
        .enumerate()
        .map(|(i, header_value)| (i + 1).to_string() + ". " + &parse_receive_header(header_value))
        .collect::<Vec<_>>()
        .join("\n")
```

Reply:
So, essentially, you don't need the `collect::<Vec<_>>()` and `.iter().map(|a| a.to_string())`.